pulley/src/lib.rs

```diff
@@ -15,6 +15,68 @@ extern crate std;
 extern crate alloc;
 
 /// Calls the given macro with each opcode.
+///
+/// # Instruction Guidelines
+///
+/// We're inventing an instruction set here which naturally brings a whole set
+/// of design questions. Note that this is explicitly intended to be only ever
+/// used for Pulley where there are a different set of design constraints than
+/// other instruction sets (e.g. general-purpose CPU ISAs). Some examples of
+/// constraints for Pulley are:
+///
+/// * Instructions must be portable to many architectures.
+/// * The Pulley ISA is mostly target-independent as the compilation target is
+///   currently only parameterized on pointer width and endianness.
+/// * Pulley instructions must be fast to decode at runtime (e.g. it's theorized
+///   we can't get super fancy with bit-packing)
+/// * Many "macro ops" are present to reduce the number of opcodes so there is a
+///   wide set of duplicate functionality between opcodes (and this is
+///   expected).
+///
+/// Given all this it's also useful to have a set of guidelines used to name and
+/// develop Pulley instructions. As of the time of this writing it's still
+/// pretty early days for Pulley so some of these guidelines may change over
+/// time. Additionally instructions don't necessarily all follow these
+/// conventions and that may also change over time. With that in mind, here's a
+/// rough set of guidelines:
+///
+/// * Most instructions are prefixed with `x`, `f`, or `v`, indicating which
+///   type of register they're operating on. (e.g. `xadd32` operates on the `x`
+///   integer registers and `fadd32` operates on the `f` float registers).
+///
+/// * Most instructions are suffixed or otherwise contain the bit width they're
+///   operating on. For example `xadd32` is a 32-bit addition.
+///
+/// * Instructions operate on either 32 or 64-bit parts of a register.
+///   Instructions modifying only 32-bits of a register always modify the "low"
+///   part of a register and leave the upper part unmodified. This is intended
+///   to help 32-bit platforms where if most operations are 32-bit there's no
+///   need for extra instructions to sign or zero extend and modify the upper
+///   half of the register.
```

Comment:
Curious on your thoughts on this in particular @fitzgen.

I'll note that we don't currently do this well, e.g. `xconst8` always writes the full 64-bits and `br_if` has no variant that only reads 32-bits. I think we'll want to change that over time though.

Reply:
This I feel is a bit entangled with the question below of "do we want `xload8_s32_offset32`". I'm a bit worried about `xconst8` always filling the entire register mostly from the perspective of 32-bit platforms. If most constants in a program are actually 32-bits then we're in theory spending a lot of instructions to fill the upper half of the register that's never read. Cache-wise that's probably not too problematic but it's possibly problematic from an icache perspective of the implementation of `xconst8` itself.

I definitely agree though that for 64-bit platforms it doesn't matter at all.

For now I'm going to go ahead with this PR as-is (with `xload8_s32_offset32` for example) and these docs since we don't have a great answer one way or another and I think we're both on the fence. I'm going to add some notes to https://github.com/bytecodealliance/wasmtime/issues/9747 to come back to reevaluate though.